javaparser-core/src/main/java/com/github/javaparser/utils/SourceZip.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * JavaParser can be used either under the terms of
+ * a) the GNU Lesser General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ * b) the terms of the Apache License
+ *
+ * You should have received a copy of both licenses in LICENCE.LGPL and
+ * LICENCE.APACHE. Please refer to those files for details.
+ *
+ * JavaParser is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+package com.github.javaparser.utils;
+
+import static com.github.javaparser.ParseStart.COMPILATION_UNIT;
+import static com.github.javaparser.Providers.provider;
+import static com.github.javaparser.utils.Utils.assertNotNull;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.ParseResult;
+import com.github.javaparser.ParserConfiguration;
+import com.github.javaparser.ast.CompilationUnit;
+
+/**
+ * A collection of Java source files and its sub-directories located in a ZIP or JAR file on the file system.
+ * Files can be parsed with a callback.
+ */
+public class SourceZip {
+
+    private final Path zipPath;
+    private JavaParser javaParser;
+
+    /**
+     * Create a new ZIP parser. An instance of {@link JavaParser} with the default {@link ParserConfiguration} will be
+     * used to parse the ZIP.
+     *
+     * @param zipPath The absolute path of ZIP file to parse.
+     */
+    public SourceZip(Path zipPath) {
+        this(zipPath, new ParserConfiguration());
+    }
+
+    /**
+     * Create a new ZIP parser. An instance of {@link JavaParser} with the given configuration will be used to parse
+     * the ZIP.
+     *
+     * @param zipPath The absolute path of ZIP file to parse.
+     * @param configuration The configuration to initiate the default parser with.
+     */
+    public SourceZip(Path zipPath, ParserConfiguration configuration) {
+        assertNotNull(zipPath);
+        assertNotNull(configuration);
+        this.zipPath = zipPath.normalize();
+        this.javaParser = new JavaParser(configuration);
+        Log.info("New source zip at \"%s\"", this.zipPath);
+    }
+
+    /**
+     * Tries to parse all '.java' files in the ZIP located at this <i>SourceZip</i>'s path and passes them one by one
+     * to the provided callback.
+     *
+     * @param callback A callback method for the post-processing of a parsed ZIP file.
+     * @return This {@link SourceZip} object.
+     *
+     * @throws {@link IOException} If an error occurs while trying to parse the given source.
+     */
+    public SourceZip parse(Callback callback) throws IOException {

```

Comment:
Should we also offer a simpler facade that just call this method internally and give you back a list of ASTs or maybe a list of pairs Path-AST?

Reply:
We decided in the issue to keep it to this for now. I guess it can be added easily. @ryan-beckett can decide if he wants to add this or not :-)

Reply:
@ftomassetti, @matozoid : I think that's better. I'll work on it when I get home.

Reply:
@ftomassetti : When you say facade, do you mean simply change the signature to:

`public List<ParseResult<CompilationUnit>> parse()`

or are we talking about something more in depth?

Reply:
@matozoid, @ftomassetti: I've committed to the PR. Let me know you think.